README.md
<img alt="bob" align="left" src="K8-Beeline.png" width="20%" height="20%">

## beeline.k8s - A library of Kubernetes Shortcuts

This is a collection of kubectl and bash/zsh shortcuts to make interactions with Kubernetes a bit easier.

![Hex.pm](https://img.shields.io/hexpm/l/apa)
[![CodeFactor](https://www.codefactor.io/repository/github/262life/beeline/badge)](https://www.codefactor.io/repository/github/262life/beeline)
![GitHub Workflow Status (branch)](https://img.shields.io/github/workflow/status/262life/beeline/release/v2.2.5?label=build%20v2.2.5)
![GitHub release (latest SemVer)](https://img.shields.io/github/v/release/262life/beeline)
![Twitter Follow](https://img.shields.io/twitter/follow/262life?style=social)
[![Gitpod ready-to-code](https://img.shields.io/badge/Gitpod-ready--to--code-blue?logo=gitpod)](https://gitpod.io/#https://github.com/262life/beeline)

beeline.k8s integrates into your bash or zsh shell to provide a number of useful shortcuts as well as the ability to keep individual kubectl config files for each cluster.  The makes mangaging the configs much easier to read and manage in the end.

For more details about the solutions currently supported by beeline.k8s, please refer to the [project status section](#project-status) below.
We plan to continue adding support for many common functions required based on community demand and engagement in future releases. 

## Getting Started 

To use the shortcut scripts it is *required* that you have setup either bash or zsh completion in advance.  Beeline will enable kubectl's auto completion on your behalf.  Also, you must have installed kubectl and helm both in /usr/local/bin and they must be executable.

### Installing ZSH 
This has been tested on macOS, Ubuntu, and Windows WSL2.  

I highly recommend changing your login shell to zsh.  It is *now* the <span style="color:lightgreen">*macOS*</span> default.  For Ubuntu or WSL2, the following will install zsh.  If prompted, choose any defaults.

```
sudo apt install zsh
```

and to enable autocomplete for zsh on aly of the supported environments, install oh-my-zsh as follows:

```
sh -c "$(curl -fsSL https://raw.github.com/ohmyzsh/ohmyzsh/master/tools/install.sh)"
```

To enable these shortcuts you must source them into your shell.  This is dependent on the shell you are using.
You can either download the desired version you would like or if you are lazy like me do something like this if you use zsh:

```
#####  Setup Beeline

## Defaults if you like
export KS_CONTEXT='cities'
export KS_NAMESPACE='kube-system'

[[ ! -f ~/.beeline.k8s ]]  \
&& curl -s -L https://github.com/262life/beeline/releases/latest/download/beeline.sh  > ~/.beeline.k8s

source ~/.beeline.k8s
##### End of Beeline
```
Take note to edit the KS_CONTEXT and KS_NAMESPACE values to your liking.
Restart your shell and you should be good to go!  

Now, review the [usage documentation](DOCUMENTATION.md) 

## NOTES

- This is an entirely new concept that allows a different context/namespace in each terminal window.  It supports multiple logins too...
- Significantly improves the performance
- Currently user should ONLY use zsh for your shell for the terminal.  BASH is fine for scripting.

### About Kubectl config files

We made an improvement to kubectl config files to allow you to keep seperate kubectl files for each kubeernetes environment.  Just create a single config file for each cluster and name it some clustername.cfg and put it in the kube folder.  You can look in the examples folder for ideas.  Please note that the tokens and certs are intentionlly non working so replace with your own of course. 

## Contributing

We welcome contributions. 

## Report a Bug

For filing bugs, suggesting improvements, or requesting new features, please open an [issue](https://github.com/262life/beeline.k8s/issues).

## Contact

Please use the following to reach members of the community:

- Email: [support@262.life](mailto:support@262.life)

## Security

### Reporting Security Vulnerabilities

If you find a vulnerability or a potential vulnerability in beeline.k8s please let us know immediately at
[support@262.life](mailto:support@262.life). We'll send a confirmation email to acknowledge your
report, and we'll send an additional email when we've identified the issues positively or
negatively.


## Project Status

We consider the latest release to be **stable**.  If you use a release candidate proceed with caution. 


### Official Releases

Official releases of beeline.k8s can be found on the [releases page](https://github.com/262life/beeline/releases).
Please note that it is **strongly recommended** that you use [official releases](https://github.com/262life/beeline/releases) of beeline.k8s, as unreleased versions from the master branch are subject to changes and incompatibilities that will not be supported in the official releases.
Builds from the master branch can have functionality changed and even removed at any time without compatibility support and without prior notice.

## Licensing

beeline.k8s is under the Apache 2.0 [License.](LICENSE)



CHANGES.md

## [2.2.5] - 2022-12-24
### Changes

- Improved functions if no default KS_CONTEXT is set
- Added a check for .kube folder
- Added a test for empty properties file
- Fixed if running in a script

## [2.2.4] - 2022-09-23
### Add

-  Added a new ku function to force an update check

## [2.2.3] - 2022-09-23
### Changes

-  One fix for kc command for PS1 variable visiability
-  Untangled some code around kc and kn command
-  Suppressed some context output
-  Removed some useless cat commands based on spellcheck recommendations

## [2.2.2] - 2022-05-27
### Changes

-  Small fix for kc command

## [2.2.1] - 2022-05-10
### Changes

-  Small fix for XTerminals

## [2.2.0] - 2022-05-05
### Changes

-  Incorporating update check logic
-  New kv shortcut to show running version

## [2.2.0-RC2] - 2022-05-05
### Changes

-  Incorporating update check logic

## [2.2.0-RC1] - 2022-04-17
### Changes

-  Getting ready for next release

## [2.1.2] - 2021-11-7
### Changes

- Small Fixes

## [2.1.1] - 2021-09-14
### Changes

- Added new utility function to check for the required dependencies
- improved the documentation

## [2.1.0] - 2021-08-26
### Changes

- Added new utility functions to create a list
- improved the documentation

## [2.0.0] - 2021-08-14
### Changes

- This is an entirely new concept that allows a different context/namespace in each terminal window.  It supports multiple logins too...
- Significantly improved the performance
- Currently use should ONLY use zsh for your shell for the terminal.  BASH is fine for scripting

## [2.0.0-RC1] - 2021-07-03
### Changes

- This is an entirely new concept that allows a different context/namespace in each terminal window.  It supports multiple logins too...

## [1.0.1] - 2021-03-21
### Changes

- Mahendran Mookkiah provided a better method for kn() function

## [1.0.0] - 2020-07-31
### Changes

- Initial Release





tu.sh


check_for_update() {

  if [[ $(find ${ok_file} -mtime +${days}) = "$ok_file" ]]; then
    candidate=$(curl -s -L https://github.com/262life/beeline/releases/${release}/checksum.sha512)
    current=$(cat ~/.beeline.k8s | shasum -a 512)
    if [[ "$candidate" != "$current" ]] ; then

      # Ask for confirmation and only update on 'y', 'Y' or Enter
      # Otherwise just show a reminder for how to update
      echo -n "${cyan}[beeline] Would you like to update? ${white}[${cyan}Y/n${white}]${defcolor} "
      read -r -k 1 option
      [[ "$option" = $'\n' ]] || echo
      case "$option" in
        [yY$'\n']) update_beeline ;;
        [nN]) update_last_updated_file ;&
        *) echo "[beeline] You can update manually by ........" ;;
      esac
    fi
  fi

}

update_last_updated_file() {

  touch  ${ok_file}

}

update_beeline() {

  curl -s -L https://github.com/262life/beeline/releases/${release}/beeline.sh > "${HOME}/.beeline.k8s"
  update_last_updated_file
  source  ${HOME}/.beeline.k8s

}




beeline (conflicted copy 2022-04-17 131710).sh
#!/bin/bash      
### ^^^ Needed for Shellcheck.  Not intended to run this as a script.  It must be sourced.


### Main Function - Setup goes here
function main() {

  [[ -t fd  ]] && export TERMINAL=true  # Check if this is an interactive session
  
  if [ "$TERMINAL" = true ] && [ $SHLVL = 1 ]; then
    case $SHELL in
      /bin/zsh )   unsetopt complete_aliases; export KS_SHELL=zsh;;
      /bin/bash)  complete -F _complete_alias; export KS_SHELL=bash;;
    esac
  fi

  # shellcheck disable=SC1090
  source <(/usr/local/bin/kubectl completion "$KS_SHELL")


  # set the default context and namespace from the .zshrc file
  if [ $SHLVL = 1 ] || [ "$TERM_PROGRAM" = "vscode" ]; then
  kc >/dev/null
  fi

}

### Functions to match aliases -  These are used for the shell scripts and the aliases are needed to autocompletion

function kubectl()      { NO_PROXY="" HTTPS_PROXY="" no_proxy="" https_proxy="" HTTP_PROXY="" http_proxy="" /usr/local/bin/kubectl "$@"; } 
function k()            { kubectl "$@"; }
function kd()           { kubectl describe "$@"; }
function kdp()          { kubectl describe pod "$@";}
function kds()          { kubectl describe service "$@"; }
function ke()           { kubectl edit "$@";}
function kh()           { khelp "$@"; }
function kg()           { kubectl get "$@"; }
function kgp()          { kubectl get pods "$@";}
function kgs()          { kubectl get services "$@"; }
function kl()           { kubectl logs "$@"; }
function kn()           { if [ -n "$1" ]; then export KS_NAMESPACE=${1}; fi; kc ""; }

### Utility Functions

function kfl()          { [[ "$#" -gt 1 ]] && { kf "$@" | awk '{print $1}' | grep -v 'NAME' | xargs; } || echo "kfl requires exactly 2 parameters"; }
function kf()           { [[ "$#" -gt 1 ]] && { O=$1; shift; F="${*}"; k get "${O}" | grep -E "NAME|$F"; } || echo "kf requires exactly 2 parameters"; }

### Other miscellaneous aliases
function helm()         { /usr/local/bin/helm "$@"; }
function hs()           { fc -lim "*${*}*" 1; }
function kubeconfig()   {   export KUBECONFIG; KUBECONFIG="$(  (ls ~/.kube/*.cfg; ls ~/.kube/config ) | xargs | sed -e "s/ /:/g")"; echo "$KUBECONFIG"; }
function velero()       { /usr/local/bin/velero "$@"; }

function kc() {
  
  if [ -z "${1}" ]; then
    echo ""
    [[ $TERMINAL = true ]] && [ $SHLVL = 1 ] && set -o PROMPT_SUBST
    cyan=$(printf     '\e[0m\e[36m')
    white=$(printf    '\e[0m\e[97m')
    defcolor=$(printf '\e[0m\e[39m')
    export PS1="%{${white}%}[B:%{${cyan}%}$KS_CONTEXT:$KS_NAMESPACE%{${white}%}]%{${defcolor}%} %n:%/$ "
    KS_CLUSTER=$(/usr/local/bin/kubectl get --context "$KS_CONTEXT" cn -o custom-columns=:.spec.clusterName --no-headers=true 2>/dev/null) 
    [[ -n "$KS_CLUSTER" ]] && echo "Cluster    : $KS_CLUSTER";
    echo "Context    : $KS_CONTEXT";
    echo "Namespace  : $KS_NAMESPACE";
    echo ""
    echo ""
  else
    export KS_CONTEXT=${1}
    if [ -z "${2}" ]; then kc ""; else kn "${2}"; fi
  fi


  export KUBECONFIG; KUBECONFIG=$(kubeconfig)
  /usr/local/bin/kubectl config view --context "$KS_CONTEXT" --namespace "$KS_NAMESPACE" --minify --raw=true -oyaml > "$HOME/.kube/beeline.properties.$$" \
    && chmod 700 "$HOME/.kube/beeline.properties.$$"
  
  export KUBECONFIG; KUBECONFIG="$HOME/.kube/beeline.properties.$$"
  /usr/local/bin/kubectl config set-context "$KS_CONTEXT" --namespace="$KS_NAMESPACE" 
  /usr/local/bin/kubectl config set current-context "$KS_CONTEXT" 
 
  # reset aliases

  alias kubectl="NO_PROXY='' HTTPS_PROXY='' no_proxy='' https_proxy='' HTTP_PROXY='' http_proxy='' /usr/local/bin/kubectl"
  alias k="kubectl"
  alias kd="kubectl describe"
  alias kdp="kubectl describe pod"
  alias kds="kubectl describe service"
  alias ke="kubectl edit"
  alias kh="khelp"
  alias kg="kubectl get"
  alias kgp="kubectl get pods"
  alias kgs="kubectl get services"
  alias kl="kubectl logs"

  ### Other miscellaneous aliases
  alias helm="/usr/local/bin/helm"
  alias velero="/usr/local/bin/velero"


  if [ -z "${2}" ]; then 
    echo ""
    /usr/local/bin/kubectl get --context "$KS_CONTEXT" nodes -o custom-columns='NAME:.metadata.name,STATUS:.status.conditions[?(@.reason=="KubeletReady")].type,ARCH:.status.nodeInfo.architecture,INTERNAL ADDRESS:status.addresses[?(@.type=="InternalIP")].address,KERNEL:.status.nodeInfo.kernelVersion,VERSION:.status.nodeInfo.kubeletVersion,OS:.status.nodeInfo.osImage,CONTAINER RUNTIME:.status.nodeInfo.containerRuntimeVersion' | colorize
  fi

}


khelp() {

cat <<EOD
  beeline - Version: v0.0.0

  These are the latest shortcuts supported.  You will find autocomplete works on all.

  kubectl = improved kubectl with seperate configs
  k       = kubectl
  kd      = kubectl describe
  kdp     = kubectl describe pod
  kds     = kubectl describe service
  ke      = kubectl edit
  kh      = this help page
  kg      = kubectl get
  kgp     = kubectl get pods
  kgs     = kubectl get services
  kl      = kubectl logs

  helm    = improved helm v3 with seperate configs
  velero  = improved velero with seperate configs

  Special utility functions:

  kf      = kubectl find.  This will the current context and filter the list.  You can use a '|' character to concatenate 
            more than one filter.  Example:  Assuming the cluster is set and the namespace is kube-system *kf pod "coredns|calico" 
            will return all pods that include those strings.
  kfl     = Same as *kf* but will return a list appropriate to include as a list in a kubectl command
            Example:  *kgp $(kfl service "coredns|calico")* will return a filtered list of pods.  This can be used with all shortcuts.

EOD

}

function checks() {

  let bok=0
  [[ ! -f ~/.beeline.ok ]] \
     && { echo "Beeline - pre launch check... \n";
          [[ -z "${ZSH_VERSION}" ]] \
             && { echo -n "oh-my-zsh not installed.  "
                  echo "Please install. Refer to "https://github.com/ohmyzsh/ohmyzsh" for help" 
                  let bok+=1
                }
          [[ ! -x "/usr/local/bin/helm" ]] \
             && { echo -n "helm not installed.  "
                  echo "Please install in /usr/local/bin. Refer to "https://helm.sh/docs/intro/install/" for help"
                  let bok+=1  
                }
          [[ ! -x "/usr/local/bin/kubectl" ]] \
             && { echo -n "kubectl not installed.  "
                  echo "Please install in /usr/local/bin. Refer to "https://kubernetes.io/docs/tasks/tools/#kubectl" for help" 
                  let bok+=1 
                }
        }  
  [[ $bok -eq 0 ]] \
     && { touch ~/.beeline.ok 
          return 0
        } || return $bok

}

function colorize() {
  awk '
  function color(c,s) {
           printf("\033[%dm%s\033[0m\n",30+c,s)
  }
  /amd64/ {color(7,$0);next}
  /arm64/ {color(5,$0);next}
  {print}
  ' $1

}

#########################  MAIN Script starts here ###############################

# Version: v0.0.0

# shellcheck disable=SC2064
trap "rm -f ${HOME}/.kube/beeline.properties.${$}" EXIT
checks && main "$@"


beeline.sh
#!/bin/bash      
### ^^^ Needed for Shellcheck.  Not intended to run this as a script.  It must be sourced.

### Main Function - Setup goes here
function main() {

  [[ -t fd  ]] && export TERMINAL=true  # Check if this is an interactive session
  if [ $SHLVL = 1 ] || [ $SHLVL = 2 ];then  export BEELINE=true; fi

  if [ "$TERMINAL" = true ] && [ "$BEELINE" = true ]; then
    case $SHELL in
      /bin/zsh )   unsetopt complete_aliases; export KS_SHELL=zsh;;
    esac
  fi

  # shellcheck disable=SC1090
    source <(/usr/local/bin/kubectl completion "$KS_SHELL") 2>/dev/null


  # set the default context and namespace from the .zshrc file
  if [ "$BEELINE" = true ] || [ "$TERM_PROGRAM" = "vscode" ]; then
    kc >/dev/null
  fi

}

### Functions to match aliases -  These are used for the shell scripts and the aliases are needed to autocompletion

function kubectl()      { NO_PROXY="" HTTPS_PROXY="" no_proxy="" https_proxy="" HTTP_PROXY="" http_proxy="" /usr/local/bin/kubectl "$@"; } 
function k()            { kubectl "$@"; }
function kd()           { kubectl describe "$@"; }
function kdp()          { kubectl describe pod "$@";}
function kds()          { kubectl describe service "$@"; }
function ke()           { kubectl edit "$@";}
function kh()           { khelp "$@"; }
function kg()           { kubectl get "$@"; }
function kgp()          { kubectl get pods "$@";}
function kgs()          { kubectl get services "$@"; }
function kl()           { kubectl logs "$@"; }
function kn()           { if [ -n "$1" ]; then export KS_NAMESPACE=${1}; fi; kc ""; }
function ku()           { touch -d 2000-01-01T00:00:00 ~/.beeline.ok; check_for_update; }
function kv()           { echo "kubectl:"; kubectl version; echo "beeline:"; grep '^# Version: ' "${HOME}/.beeline.k8s" | sed -e 's/^# //g'; }

### Utility Functions

function kfl()          { [[ "$#" -gt 1 ]] && { kf "$@" | awk '{print $1}' | grep -v 'NAME' | xargs; } || echo "kfl requires exactly 2 parameters"; }
function kf()           { [[ "$#" -gt 1 ]] && { O=$1; shift; F="${*}"; k get "${O}" | grep -E "NAME|$F"; } || echo "kf requires exactly 2 parameters"; }

### Other miscellaneous aliases
function helm()         { /usr/local/bin/helm "$@"; }
function hs()           { fc -lim "*${*}*" 1; }
function kubeconfig()   {   export KUBECONFIG; KUBECONFIG="$(  (ls ~/.kube/*.cfg; ls ~/.kube/config ) 2>/dev/null | xargs | sed -e "s/ /:/g")"; echo "$KUBECONFIG"; }
function velero()       { /usr/local/bin/velero "$@"; }

function kc() {

  if [[ -z "${1}" ]] && [[ ! -z "$KS_CONTEXT" ]]; then
    echo ""
    [[ $TERMINAL = true ]] && [ $SHLVL = 1 ] && set -o PROMPT_SUBST
    KS_CLUSTER=$(/usr/local/bin/kubectl get --context "$KS_CONTEXT" cn -o custom-columns=:.spec.clusterName --no-headers=true 2>/dev/null ) 
    [[ -n "$KS_CLUSTER" ]] && echo "Cluster    : $KS_CLUSTER";
    echo "Context    : $KS_CONTEXT";
    echo "Namespace  : $KS_NAMESPACE";
  else
    export KS_CONTEXT=${1}
    if [[ -n "${2}" ]] && [[ ! -z "$KS_CONTEXT" ]]; then kn "${2}"; fi
  fi

  export PS1="%{${white}%}[B:%{${cyan}%}$KS_CONTEXT:$KS_NAMESPACE%{${white}%}]%{${defcolor}%} %{${cyan}%}%n@%m%{${defcolor}%}:%/$ "

  export KUBECONFIG; KUBECONFIG=$(kubeconfig)

  if [[ ! -z "$KS_CONTEXT" ]]; then
    /usr/local/bin/kubectl config view --context "$KS_CONTEXT" --namespace "$KS_NAMESPACE" --minify --raw=true -oyaml  > "$HOME/.kube/beeline.properties.$$" \
      && chmod 700 "$HOME/.kube/beeline.properties.$$"

    export KUBECONFIG; KUBECONFIG="$HOME/.kube/beeline.properties.$$"
    /usr/local/bin/kubectl config set-context "$KS_CONTEXT" --namespace="$KS_NAMESPACE"  >/dev/null
    /usr/local/bin/kubectl config set current-context "$KS_CONTEXT" >/dev/null
  fi
 
  # reset aliases

  alias kubectl="NO_PROXY='' HTTPS_PROXY='' no_proxy='' https_proxy='' HTTP_PROXY='' http_proxy='' /usr/local/bin/kubectl"
  alias k="kubectl"
  alias kd="kubectl describe"
  alias kdp="kubectl describe pod"
  alias kds="kubectl describe service"
  alias ke="kubectl edit"
  alias kh="khelp"
  alias kg="kubectl get"
  alias kgp="kubectl get pods"
  alias kgs="kubectl get services"
  alias kl="kubectl logs"

  ### Other miscellaneous aliases
  alias helm="/usr/local/bin/helm"
  alias velero="/usr/local/bin/velero"

  if [[ -z "${2}" ]] &&  [[ ! -z "$KS_CONTEXT" ]]; then
    echo ""
    /usr/local/bin/kubectl get --context "$KS_CONTEXT" nodes -o custom-columns='NAME:.metadata.name,STATUS:.status.conditions[?(@.reason=="KubeletReady")].type,ARCH:.status.nodeInfo.architecture,INTERNAL ADDRESS:status.addresses[?(@.type=="InternalIP")].address,KERNEL:.status.nodeInfo.kernelVersion,VERSION:.status.nodeInfo.kubeletVersion,OS:.status.nodeInfo.osImage,CONTAINER RUNTIME:.status.nodeInfo.containerRuntimeVersion' | colorize
  fi

}

function khelp() {

cat <<:EOD
  beeline - Version: v2.2.4

  These are the latest shortcuts supported.  You will find autocomplete works on all.

  kubectl = improved kubectl with seperate configs
  k       = kubectl
  kd      = kubectl describe
  kdp     = kubectl describe pod
  kds     = kubectl describe service
  ke      = kubectl edit
  kh      = this help page
  kg      = kubectl get
  kgp     = kubectl get pods
  kgs     = kubectl get services
  kl      = kubectl logs
  ku      = force an update check for beeline
  kv      = kubectl & beeline version

  helm    = improved helm v3 with seperate configs
  velero  = improved velero with seperate configs

  Special utility functions:

  kf      = kubectl find.  This will the current context and filter the list.  You can use a '|' character to concatenate 
            more than one filter.  Example:  Assuming the cluster is set and the namespace is kube-system *kf pod "coredns|calico" 
            will return all pods that include those strings.
  kfl     = Same as *kf* but will return a list appropriate to include as a list in a kubectl command
            Example:  *kgp \$(kfl service "coredns|calico")* will return a filtered list of pods.  This can be used with all shortcuts.

:EOD

}

function checks() {

  (( bok=0 ))
  [[ ! -f ${ok_file} ]] \
     && { printf "Beeline - pre launch check... \n";
          [[ -z "${ZSH_VERSION}" ]] \
             && { echo -n "beeline not installed.  "
                  echo "Please install. Refer to 'https://github.com/262life/beeline' for help" 
                  (( bok+=1 ))
                }
          [[ ! -x "/usr/local/bin/helm" ]] \
             && { echo -n "helm not installed.  "
                  echo "Please install in /usr/local/bin. Refer to 'https://helm.sh/docs/intro/install/' for help"
                  ((bok+=1))  
                }
          [[ ! -x "/usr/local/bin/kubectl" ]] \
             && { echo -n "kubectl not installed.  "
                  echo "Please install in /usr/local/bin. Refer to 'https://kubernetes.io/docs/tasks/tools/#kubectl' for help" 
                  (( bok+=1 ))
                }
          [[ ! -d "$HOME/.kube" ]] \
             && { mkdir "$HOME/.kube"; touch "$HOME/.kube/config" }
        }  
  [[ $bok -eq 0 ]] \
     && { touch "${ok_file}"
          return 0
        } || return $bok

}

function colorize() {
  awk '
  function color(c,s) {
           printf("\033[%dm%s\033[0m\n",30+c,s)
  }
  /amd64/ {color(7,$0);next}
  /arm64/ {color(5,$0);next}
  {print}
  '   #### $1

}

check_for_update() {

  if [[ $(find "${ok_file}" -mtime +"${days}" 2>/dev/null) = "$ok_file" ]]; then
    candidate=$(curl -s -L "https://github.com/262life/beeline/releases/${release}/checksum.sha512")
    current=$(shasum -a 512 < ~/.beeline.k8s)
    if [[ "$candidate" != "$current" ]] ; then

      # Ask for confirmation and only update on 'y', 'Y' or Enter
      # Otherwise just show a reminder for how to update
      echo -n "${cyan}[beeline] Would you like to update? ${white}[${cyan}Y/n${white}]${defcolor} "
      read -r -k 1 option
      [[ "$option" = $'\n' ]] || echo;
      case "$option" in
        [yY$'\n']) update_beeline ;;
        [nN]) update_last_updated_file ;;
        *) echo "[beeline] You can update manually by ........" ;;
      esac
    fi
  fi

}

update_last_updated_file() {

  touch  "${ok_file}"

}

update_beeline() {
 
  echo  "${cyan}Updating...${defcolor} "
  # shellcheck disable=SC1091
  curl -s -L "https://github.com/262life/beeline/releases/${release}/beeline.sh" > "${HOME}/.beeline.k8sX"  && \
  mv "${HOME}/.beeline.k8sX" "${HOME}/.beeline.k8s"  && \
  update_last_updated_file && \
  source "${HOME}/.zshrc" || \
  echo "${cyan}Updating failed...${defcolor} "

}


#########################  MAIN Script starts here ###############################

# Version: v2.2.4

# Global settings
release="latest/download"  #<-- to test: release="download/v2.2.0-RC1"
ok_file="${HOME}/.beeline.ok"
days=7

# Check for build
if [[ "${1}" = "--help" ]]; then 
  khelp 
  exit 
fi;

# Set cool colors
cyan=$(printf     '\e[0m\e[36m')
white=$(printf    '\e[0m\e[97m')
defcolor=$(printf '\e[0m\e[39m')

# shellcheck disable=SC2064
trap "rm -f ${HOME}/.kube/beeline.properties.${$} 2>/dev/null" EXIT
check_for_update
checks && main "$@"
